# Estudio de cadenas de Markov 

https://d3c33hcgiwev3.cloudfront.net/_adadc80290e52a99b282ca9d7c1a41ee_background_MarkovChains.html?Expires=1609632000&Signature=ULgkm0IDwpxNss5xd9-cueDURKcq6sU4GEnhsi-OwsMyVRj65AY4ytrZVsqqOcIeT~5yp9bFq2qoqb~x3ISnYaLzMmEDOsjsUppu30JjfzHny0klZ5LVvz8lVsfCqX2ypySp-GNREepb8DDPFnJJw3e3JzmIDoo5xEpcG6Uljto_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A

In [1]:
from scipy.stats import beta,gamma,norm,invgamma,uniform
import scipy.special
import numpy as np
from sklearn.linear_model import LinearRegression
from scipy import optimize

In [2]:
import matplotlib.pyplot as plt

In [3]:
import pandas as pd
import math

In [4]:
import arviz as az

import pymc3 as pm

from pymc3 import *

import theano.tensor as tt
import warnings;
warnings.filterwarnings('ignore');

In [5]:
pm.__version__

'3.8'

In [17]:
Q=np.array([[0.0, 0.5, 0.0, 0.0, 0.5],
             [0.5, 0.0, 0.5, 0.0, 0.0],
             [0.0, 0.5, 0.0, 0.5, 0.0],
             [0.0, 0.0, 0.5, 0.0, 0.5],
             [0.5, 0.0, 0.0, 0.5, 0.0]])

In [18]:
Q_2=np.matmul(Q,Q)

In [19]:
Q_2[1,3]

0.25

In [20]:
Q_5 = np.linalg.matrix_power(Q, 5)

In [21]:
Q_5

array([[0.0625 , 0.3125 , 0.15625, 0.15625, 0.3125 ],
       [0.3125 , 0.0625 , 0.3125 , 0.15625, 0.15625],
       [0.15625, 0.3125 , 0.0625 , 0.3125 , 0.15625],
       [0.15625, 0.15625, 0.3125 , 0.0625 , 0.3125 ],
       [0.3125 , 0.15625, 0.15625, 0.3125 , 0.0625 ]])

Veamos si tiene estado estacionario

In [140]:
n=5000
parrilla=np.array([1,2,3,4,5])

In [141]:
x=np.zeros(shape=n).astype(int)

In [142]:
x.shape

(5000,)

In [143]:
x[0]=1

In [144]:
for i in range(1,n):
    x[i] = np.random.choice(parrilla,size=1,p=Q[x[i-1]-1,])

In [145]:
np.count_nonzero(x==1)

959

In [146]:
unique, counts = np.unique(x, return_counts=True)

In [147]:
print (np.asarray((unique,counts)).T)

[[   1  959]
 [   2 1040]
 [   3 1070]
 [   4 1000]
 [   5  931]]


# Test

In [149]:
matriz=np.array([[0,1],[0.3,.7]])

In [164]:
matriz_5 = np.linalg.matrix_power(matriz, 4)

In [165]:
matriz_5

array([[0.237 , 0.763 ],
       [0.2289, 0.7711]])

In [163]:
np.matmul(np.matmul(np.matmul([1,0],matriz),matriz),matriz)

array([0.21, 0.79])

In [170]:
np.matmul([1,0],np.linalg.matrix_power(matriz, 3))

array([0.21, 0.79])

La pobilidad de que juegue la cuarta partida no jugando la primera es de 0.79